### PR TITLE
Telink B91 ADC driver support

### PR DESCRIPTION
**Pull request by yurvyn**

### CODE REVIEW

**Comment by yurvyn** [comment]
@anangl, @carlescufi, @mbolivar-nordic
Please provide your comments and approve the PR.

**Comment by yurvyn** [comment]
Hello @anangl 
Please approve the PR.

File: drivers/adc/adc_b91.c
Comment (fkokosinski) [inline drivers/adc/adc_b91.c:21]: Use `BIT*` macros here.

File: drivers/adc/adc_b91.c
Comment (fkokosinski) [inline drivers/adc/adc_b91.c:120]: I'm wondering - couldn't this be somehow shortened? I.e. by using the `FOR_EACH` macro, because almost all of the cases in this switch statement seem to write the matched value to the `adc_pin` variable.

File: drivers/adc/adc_b91.c
Comment (fkokosinski) [inline drivers/adc/adc_b91.c:318]: Same here - this case should be shortened as well.

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c]: Done.

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c]: Done.

File: drivers/adc/adc_b91.c
Comment (edersondisouza) [inline drivers/adc/adc_b91.c:86]: Maybe return `status` here? Or, if the `adc_b91_validate_buffer_size` return doesn't matter, just return `bool` from it?

File: drivers/adc/adc_b91.c
Comment (anangl) [inline drivers/adc/adc_b91.c:223]: `adc_b91_acquisition_thread` is defined to take `struct device *` as parameter.

File: drivers/adc/adc_b91.c
Comment (anangl) [inline drivers/adc/adc_b91.c:120]: Since `adc_code` is a `signed short`, why not check if `adc_code < 0` instead?

File: drivers/adc/adc_b91.c
Comment (anangl) [inline drivers/adc/adc_b91.c:204]: `adc_code` is a signed integer, so when its value is negative, the result of `>>` is implementation-defined and the behavior of `<<` is undefined. I think you should rather use multiplication here.

File: samples/drivers/adc/boards/tlsr9518adk80d.overlay
Comment (anangl) [inline samples/drivers/adc/boards/tlsr9518adk80d.overlay:7]: Wouldn't it be better to include it from `dts/riscv/telink/telink_b91.dtsi` instead?

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c]: Done.

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c]: Fixed.

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c]: Fixed.

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c]: Done.

File: samples/drivers/adc/boards/tlsr9518adk80d.overlay
Comment (yurvyn) [reply samples/drivers/adc/boards/tlsr9518adk80d.overlay]: Done.

**Comment by yurvyn** [comment]
@edersondisouza @fkokosinski @anangl 
Please approve the PR.

**Review by fkokosinski — DISMISSED** [review]

File: drivers/adc/adc_b91.c
Comment (gmarull) [inline drivers/adc/adc_b91.c:372]: please don't use `DT_INST_FOREACH_STATUS_OKAY` is multiple instances are not supported. Also try to run `clang-format` on your driver, there seem to be multiple style issues.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c]: Done.
I also applied clang-format changes according to your recommendations, before that I just used checkpatch.

Comment (yurvyn) [reply drivers/adc/adc_b91.c]: Reverted back all clang-format changes because it causes a lot of Compliance Checks errors, checkpatch works much more batter in this case.

Comment (yurvyn) [reply drivers/adc/adc_b91.c]: @gmarull 
Checkpatch PASS and Compliance Checks also PASS.
Please approve the PR if you are OK with the driver.

File: drivers/adc/adc_b91.c
Comment (gmarull) [inline drivers/adc/adc_b91.c:50]: please make this part of device data

File: drivers/adc/adc_b91.c
Comment (gmarull) [inline drivers/adc/adc_b91.c:301]: please avoid instance based macros within drivers, better make use of driver's config for consistency with most drivers in tree

File: drivers/adc/adc_b91.c
Comment (gmarull) [inline drivers/adc/adc_b91.c:351]: this driver seems to support async essentially by spawning a thread that does polling, ie no IRQ/DMA based. this doesn't seem like the right approach to do "async". cc: @teburd

File: drivers/adc/adc_b91.c
Comment (gmarull) [inline drivers/adc/adc_b91.c:360]: for consistency: store values in driver's config instead.

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c:351]: This driver does not support IRQs for ADC and because of that it is implemented in the same way as other drivers without IRQ support: adc_ads1x1x.c, adc_emul.c, adc_lmp90xxx.c, adc_mcp320x.c, and adc_npcx.c.

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c:50]: @gmarull,
Sorry, I did not get the requested changes.
Could you please specify what exactly I have to change here?

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c:360]: See my previous comment (I prefer to have it in dts but not in Kconfig ..)

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c]: Done.

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c]: Done.

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c]: Done.

File: drivers/adc/adc_b91.c
Comment (gmarull) [inline drivers/adc/adc_b91.c:16]: `"adc.h"` is not a local header, but from HAL I guess. Please improve this include section, you can find some hints here: https://github.com/zephyrproject-rtos/zephyr/issues/41543

File: drivers/adc/adc_b91.c
Comment (gmarull) [reply drivers/adc/adc_b91.c]: cc: @anangl

**Comment by fkokosinski** [comment]
> @gmarull @fkokosinski @edersondisouza @anangl CI checking fails for nucleo boards, but it is not related to my changes and I do not know what to do with that. Could you please help me with this issue? Is it necessary to pass all CI checks to merge the PR?

Normally, yes - you need the CI to go green before anything gets merged. I can confirm though that I get the same warning about unused `adc_stm32_oversampling_ratioshift` on the current HEAD (3efce4c93938ebdd7d5d9b3ffdcd2b096490cea5).

**Comment by yurvyn** [comment]
> > @gmarull @fkokosinski @edersondisouza @anangl CI checking fails for nucleo boards, but it is not related to my changes and I do not know what to do with that. Could you please help me with this issue? Is it necessary to pass all CI checks to merge the PR?
> 
> Normally, yes - you need the CI to go green before anything gets merged. I can confirm though that I get the same warning about unused `adc_stm32_oversampling_ratioshift` on the current HEAD ([3efce4c](https://github.com/zephyrproject-rtos/zephyr/commit/3efce4c93938ebdd7d5d9b3ffdcd2b096490cea5)).

@fkokosinski
Thanks for the response.
Any ideas on what I can do to pass CI?
It is definitely not related to my changes, but I am afraid that it will block this PR.

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c]: Originally it was with the regular case implementation and it was changed to FOR_EACH because of the review comment in this PR.
Fixed. Reverted back to the original implementation.

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c]: Fixed.

File: dts/bindings/adc/telink,b91-adc.yaml
Comment (gmarull) [reply dts/bindings/adc/telink,b91-adc.yaml:30]: @galak @mbolivar-nordic is this a good practice?

File: dts/riscv/telink/telink_b91.dtsi
Comment (yurvyn) [reply dts/riscv/telink/telink_b91.dtsi]: Fixed, removed from DTSI, and added to the board's DTS.

File: dts/bindings/adc/telink,b91-adc.yaml
Comment (yurvyn) [reply dts/bindings/adc/telink,b91-adc.yaml:30]: @gmarull 
Since the TDS has DT_STRING_TOKEN exactly for this case, I guess it should be OK.
In another case, we do not need DT_STRING_TOKEN at all.
https://docs.zephyrproject.org/2.7.0/reference/devicetree/api.html#c.DT_STRING_TOKEN

Comment (gmarull) [reply dts/bindings/adc/telink,b91-adc.yaml:30]: The problem I see here is HAL driving DT design. Platforms that do not use HALs, or if HAL doesn't have such kind of defines, will likely do something like https://github.com/zephyrproject-rtos/zephyr/pull/46563#discussion_r910717790. I think we should be consistent. Let's wait to see DT maintainers opinions on this topic.

Comment (yurvyn) [reply dts/bindings/adc/telink,b91-adc.yaml:30]: @gmarull 
Other vendors also use this approach: e.g. `clock-source` in `nordic,nrf-pdm.yaml`.

Comment (yurvyn) [reply dts/bindings/adc/telink,b91-adc.yaml:30]: @gmarull 
The issue with `vref-mv = <900>;` approach is that the only 2 values are exactable, so actually, it is realy `Enum` but not `Int`.
Another potential workaround is to move these parameters to the driver's KConfig and define them as enumeration as well.
But, if the current approach is exactable I prefer to leave it as it is.

**Comment by yurvyn** [comment]
@gmarull 
If you are OK with my fixes and do not have new comments, please approve the PR.

**Comment by yurvyn** [comment]
@gmarull
Please approve the PR or provide new comments.

File: include/zephyr/dt-bindings/adc/b91-adc.h
Comment (gmarull) [inline include/zephyr/dt-bindings/adc/b91-adc.h:10]: this file is not using this header, please remove it

Comment (yurvyn) [reply include/zephyr/dt-bindings/adc/b91-adc.h]: Done.

File: dts/bindings/adc/telink,b91-adc.yaml
Comment (yurvyn) [reply dts/bindings/adc/telink,b91-adc.yaml:30]: @gmarull 

Looks like it is the last unresolved comment and in order to speed up the PR, I propose the next 4 options:

1. Accept the current DT implementation. I really like it as it is more user-friendly (in my opinion) because users are informed which values are acceptable. See also all my previous comments for additional explanations why I think the current implementation should be OK.
2. Currently this PR has 14 reviewers including DT maintainers. I propose to close this comment and if somebody from DT maintainers/experts has some DT-related comments/improvements they can create new comments by themselves.
3. We can wait for DT maintainer's response, however, I do not know how much time it will take and the PR will be delayed.
4. I can also easily implement your proposal. Just please confirm it once more and I will do it.

Comment (galak) [reply dts/bindings/adc/telink,b91-adc.yaml:30]: Is there some issue with making this an enum instead of a string?

Comment (galak) [reply dts/bindings/adc/telink,b91-adc.yaml:30]: So this really should be handled as a `regulator`, but if not then I think we have already in tree usage of `vref-mv` and `zephyr,vref-mv` so we should use on of this in the binding.

Comment (yurvyn) [reply dts/bindings/adc/telink,b91-adc.yaml:30]: According to `adc-controller.yaml`, the `zephyr,vref-mv` is dedicated to the external source but not internal, this B91 SoC does not support an external source at all, so I cannot use `vref-mv`.
B91 SoC supports 2 internal sources and this is not supported by the generic `adc-controller.yaml`, and because of that I introduced this new `vref` for B91 SoC to select the internal source (ADC_VREF_0P9V or ADC_VREF_1P2V).

@galak 
Are you OK with the current `telink,b91-adc.yaml` implementation, or should I change something here?

Comment (gmarull) [reply dts/bindings/adc/telink,b91-adc.yaml:30]: The problem here is that Devicetree is being heavily influenced by the HAL. I'd expect for example to specify sample frequency in Hz, e.g. `sample-freq = <DT_FREQ_K(48)>;`, not `sample-freq = "ADC_SAMPLE_FREQ_48K";`. @galak @mbolivar-nordic opinions here?

Comment (yurvyn) [reply dts/bindings/adc/telink,b91-adc.yaml:30]: @galak, @mbolivar-nordic, @gmarull 
FYI.. My opinion is: since only 2/3 values are supported, then this is an enumeration but not int (and actually it is related more to the enumeration usage in dts)... and because of that I prefer the current implementation.

Comment (gmarull) [reply dts/bindings/adc/telink,b91-adc.yaml:30]: > @galak, @mbolivar-nordic, @gmarull FYI.. My opinion is: since only 2/3 values are supported, then this is an enumeration but not int (and actually it is related more to the enumeration usage in dts)... and because of that I prefer the current implementation.

you can still enum for `int` properties

Comment (yurvyn) [reply dts/bindings/adc/telink,b91-adc.yaml:30]: Oh, got it.. I will rework it, thanks!

**Comment by yurvyn** [comment]
@gmarull 
All comments are fixed, please approve the PR if you are OK.

File: drivers/adc/adc_b91.c
Comment (gmarull) [inline drivers/adc/adc_b91.c:19]:
```suggestion
#include <zephyr/logging/log.h>
LOG_MODULE_REGISTER(adc_b91, CONFIG_ADC_LOG_LEVEL);
```

File: dts/bindings/adc/telink,b91-adc.yaml
Comment (gmarull) [inline dts/bindings/adc/telink,b91-adc.yaml:30]: nit: please add description for these props

**Review by gmarull — COMMENTED** [review]
lgtm, please check last couple of nitpicks

File: dts/bindings/adc/telink,b91-adc.yaml
Comment (yurvyn) [reply dts/bindings/adc/telink,b91-adc.yaml]: Done.

**Comment by yurvyn** [comment]
@gmarull 
All comments are fixed, CI pass, please approve the PR if you are OK.

**Review by gmarull — DISMISSED** [review]

File: tests/drivers/adc/adc_api/src/test_adc.c
Comment (anangl) [inline tests/drivers/adc/adc_api/src/test_adc.c:106]:
```suggestion
#define ADC_DEVICE_NODE		DT_INST(0, telink_b91_adc)
```
See 7022eb05e4b68d25643e1462afcbb694a4f57bdb.

**Comment by yurvyn** [comment]
@fkokosinski @edersondisouza @anangl 
One more approval is required, please approve the PR.

File: drivers/adc/adc_b91.c
Comment (anangl) [inline drivers/adc/adc_b91.c:307]: Wouldn't a build assertion be more suitable for signaling this misconfiguration?

File: tests/drivers/adc/adc_api/src/test_adc.c
Comment (yurvyn) [reply tests/drivers/adc/adc_api/src/test_adc.c]: Done.

File: drivers/adc/adc_b91.c
Comment (yurvyn) [reply drivers/adc/adc_b91.c:307]: This case should never happen. We can not use build assertion inside the switch, I prefer to be consistent with the rest configs here and with other ADC drivers.
No changes here.

**Review by anangl — APPROVED** [review]

**Comment by yurvyn** [comment]
@gmarull @fkokosinski @edersondisouza 
The last one approval is required, please approve.

**Comment by yurvyn** [comment]
@gmarull 
Your last approval disappeared because of a new commit.
Please approve once more.

**Comment by gmarull** [comment]
> @gmarull Your last approval disappeared because of a new commit. Please approve once more.

https://www.youtube.com/watch?v=mnVNLM7ynMQ&t=1868s

**Comment by yurvyn** [comment]
@gmarull 

Since you own most of the comments, could you please approve the PR?
Thanks!

**Review by carlescufi — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

**Pull request merged**
